### ISSUE TITLE
Migrated GCM endpoints don't work

### ISSUE THREAD

**Issue by rohit-smpx** [web-push-libs/web-push#445]
**NOTE**: Please test in a least two browsers (i.e. Chrome and Firefox). This
helps with diagnosing problems quicker.

# Setup

> Please provide the following details, the more info you can provide the
> better.

*Operating System*: Linux
*Node Version*: v10.13.0
*web-push Version*: `master`

> Please select any browsers that you are experiencing problems with:

- [x] Chrome
- [ ] Firefox
- [ ] Opera for Android
- [ ] Samsung Internet Browser
- [ ] Other

Server side problem while sending pushes for migrated endpoints.

# Problem
I had a mix of old GCM endpoints and newer endpoints using vapid.

Google deprecated GCM on April 10th 2019, and according to this [migration guide](https://firebase.googleblog.com/2018/04/time-to-upgrade-from-gcm-to-fcm.html) suggested to edit all server endpoints to FCM ones (Step 4).

`web-push` library added support for FCM in #415 and it's priority is below VAPID. But this leads to the old migrated endpoints being sent with Vapid headers and failing. If it would be above Vapid then the newer endpoints registered through Vapid will fail.

# Expected

I think there should be an option to select the method to send the push with. Or allow `false` value in options for the `sendNotification` method so that I can set `vapidDetails` to `false` for the old migrated GCM(now FCM) endpoints.

# Features Used

- [x] VAPID Support
- [x] GCM API Key
- [x] Sending with Payload

**Comment by cbratschi** [comment]
We have the same issue. According to the source code, the previous GCM key is used whenever vapidDetails is not set. However, we are still seeing 401 and 403 errors with such tokens sent to the new FCM endpoints.

The old GCM endpoints seem still to work but we are getting warning mails from Google.

What is the right way to migrate the GCM tokens to FCM?

**Comment by cbratschi** [comment]
Could also mentions refresh tokens shall be supported. How are they returned?

https://developers.google.com/cloud-messaging/faq

**Comment by rohit-smpx** [comment]
> We have the same issue. According to the source code, the previous GCM key is used whenever vapidDetails is not set. However, we are still seeing 401 and 403 errors with such tokens sent to the new FCM endpoints.
> 
> The old GCM endpoints seem still to work but we are getting warning mails from Google.
> 
> What is the right way to migrate the GCM tokens to FCM?

### From what I could gather after trying everything, there are 3 kind of endpoints:
1. FCM endpoints that can be sent **only through APIKey**: These include the old **GCM** endpoints migrated to FCM
2. FCM endpoints that can be sent through VAPID keys: I think all the newer registrations are these
3. Other providers' (mozilla) VAPID endpoints

The problem is in distinguishing between the first three.

#### What I did:
- Add a `provider` column in my Endpoints table and set it to empty for all
- When sending pushes check:
  - If we don't have `auth` or `p256dh` keys (these are really old type **1**), send through FCM key
  - Otherwise
    - First try sending with FCM, if success set provider to `FCM`
    - Else try sending through VAPID and set provider to `VAPID`
- On next run use the provider column to decide which way to send push, by not setting `vapidDetails` for endpoints where provider is 'FCM'

If you were like me and have set Vapid keys using `webPush.setVapidDetails` at app start, then not setting `vapidDetails` in `sendNotification` doesn't work. So I made a fork with a minor modification that allows `vapidDetails` to be set as `null` | `false` to disable sending with VAPID. You can find it here if you need https://www.npmjs.com/package/@smpx/web-push.

I think there might be a better way than this, but this worked out for me.

**Comment by rohit-smpx** [comment]
Just remembered why I had to publish a fork. This library hasn't had a release in a long time, so even though #415 was merged, FCM support is not available in the version on npm

**Comment by marco-c** [comment]
I'll publish a new release containing #415. Does that fix your issue?

**Comment by marco-c** [comment]
0.3.4 was published.

**Comment by rohit-smpx** [comment]
> 0.3.4 was published.

Great!

> I'll publish a new release containing #415. Does that fix your issue?

Kinda. I can work with it, but it's not automatic. I still have to keep track of endpoints that have to be sent through FCM and ones that have to be sent through VAPID, and set `sendNotification` options accordingly.
Maybe a `provider` option could be added, so that it is not dependent on endpoint url parsing.
Any other ideas?

**Comment by marco-c** [comment]
>  Maybe a provider option could be added, so that it is not dependent on endpoint url parsing.

Wouldn't you still need to keep track of endpoints that have to be sent through FCM or VAPID this way? Or how do you set `provider`?

**Comment by marco-c** [comment]
> I think there should be an option to select the method to send the push with. Or allow false value in options for the sendNotification method so that I can set vapidDetails to false for the old migrated GCM(now FCM) endpoints.

You can do this with the `vapidDetails` option, you should be able to set it to null.

**Comment by marco-c** [comment]
> You can do this with the vapidDetails option, you should be able to set it to null.

Ah no, it's not feasible right now. Patches welcome! I'd accept https://github.com/smartprix/web-push/commit/660366a5399a919bb4e6211211e2d432db9d7f1d if you send a PR.

**Comment by rohit-smpx** [comment]
> > Maybe a provider option could be added, so that it is not dependent on endpoint url parsing.
> 
> Wouldn't you still need to keep track of endpoints that have to be sent through FCM or VAPID this way? Or how do you set `provider`?

Yes, that would still be needed. But a provider option would simplify the code a bit. 

My current code looks something like this:
```js
webpush.sendNotification(
  pushSubscription,
  payload,
  {
    vapidDetails: provider !== 'VAPID' ? false : undefined,
  },
);
```
But this seems a little hacky to me.

A provider option that takes either 'VAPID', 'FCM' and handles this internally seems better to me.
```js
webpush.sendNotification(
  pushSubscription,
  payload,
  {
    provider,
  },
);
```

This could just be personal preference.
I'd be happy to just open a PR with https://github.com/smartprix/web-push/commit/660366a5399a919bb4e6211211e2d432db9d7f1d

**Comment by marco-c** [comment]
I'm not a fan of `provider` because VAPID isn't really a provider, so it could be confusing. Maybe we could add an option `vapidEnabled`?

Let's do https://github.com/smartprix/web-push/commit/660366a5399a919bb4e6211211e2d432db9d7f1d first anyway.

**Comment by marco-c** [comment]
(possibly with a test, if you have time)

**Comment by rohit-smpx** [comment]
> I'm not a fan of `provider` because VAPID isn't really a provider, so it could be confusing. Maybe we could add an option `vapidEnabled`?

That's a valid point, I have been using `provider` since the beginning that I never thought if the name fits
> Let's do [smartprix@660366a](https://github.com/smartprix/web-push/commit/660366a5399a919bb4e6211211e2d432db9d7f1d) first anyway.

> (possibly with a test, if you have time)

I'll try to get it done by Monday

**Comment by rohit-smpx** [comment]
@marco-c Thanks for merging the PR. I hope there is a release soon. I'll deprecate my fork then

**Comment by marco-c** [comment]
Released 3.3.5.
